HPCC-15404 Optimize roxie standalone queries

richardkchapman opened the pull request.

richardkchapman committed:
HPCC-15404 Optimize roxie standalone queries

Step 1 - moving ownership of the threadpools into the RoxieQueue objects. This
commit should not change any behaviour.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-15404 Optimize roxie standalone queries

Start worker threads on demand.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest 
Sha: f351748d3636ddeaa9a52c19fa51d2f715194c2a
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Please review/merge

ghalliday commented on roxie/ccd/ccdqueue.cpp:
```diff
@@ -746,6 +778,11 @@ class RoxieQueue : public CInterface
             queueLength++;
             if (queueLength>maxQueueLength)
                 maxQueueLength = queueLength;
+            if (idle==0 && started < numWorkers)
```
There is a minor race condition here.  If there is 1 idle thread, and multiple enqueue()s occur at the same time, no new threads will be started.

richardkchapman replied:
I suspect avoidance of race conditions was the motivation for pre-starting the threads in the first place.

ghalliday commented on roxie/ccd/ccdqueue.cpp:
```diff
@@ -1332,6 +1371,21 @@ class CRoxieWorker : public CInterface, implements IPooledThread
+
+void RoxieQueue::abortChannel(unsigned channel)
+{
+    Owned<IPooledThreadIterator> wi = workers->running();
```
Does this need protection inside a critical section?

richardkchapman replied:
Possibly, though I suspect no more so than it did previously.

ghalliday commented:
@richardkchapman a couple of threading questions

richardkchapman committed:
HPCC-15404 Optimize roxie standalone queries

Address race condition questions, and default to prestarting threads when
deployed via ConfigMgr. Standalone Roxie programs will NOT typically load
a RoxieTopology file and will default to not prestarting threads.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest 
Sha: 2e1cb5eded075a3c9a71b457d7a453fd364a85f1
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Back to you
